tbot/machine/linux/path.py

```diff
@@ -71,13 +61,152 @@ def __init__(self, host: H, *args: typing.Any) -> None:
+                argslist[i] = arg._path
+
+        self._host: H = host
+        self._path: pathlib.PurePosixPath = pathlib.PurePosixPath(*argslist)
+
+    # tbot specific API {{{
+    # These methods are tbot-specific and do not have an equivalent in pathlib.
 
     @property
     def host(self) -> H:
         """Host associated with this path."""
         return self._host
 
+    def _local_str(self) -> str:
+        return str(self._path)
+
+    def at_host(self, host: H) -> str:
+        """
+        Convert this ``Path`` into a string representation assuming it should
+        be valid for the machine ``host``.  An exception is raised if this
+        ``Path`` is for a different machine instead.
+
+        This prevents accidentally using a ``Path`` with the wrong host as much
+        as possible.
+        """
+        if self.host != host:
+            raise Exception(f"This path ({self}) is not for the host {host!r}")
+        return str(self._path)
+
+    # }}}
+
+    # PurePosixPath like API {{{
+    # Mimick the API of pathlib's PurePosixPath as much as possible.  This
+    # allows users to use tbot's Path just like the one they know from pathlib.
+    def _prepare_args_list(self, args: Iterable[Any]) -> List[Any]:
+        ret = list()
+        for arg in args:
+            if isinstance(arg, Path):
+                assert arg.host == self.host
+                ret.append(arg._path)
+            else:
+                ret.append(arg)
+        return ret
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, Path):
+            return NotImplemented
+        return self._host == other._host and self._path == other._path
+
+    def __hash__(self) -> int:
+        return hash((self._host, self._path))
+
+    def __lt__(self, other: Any) -> bool:
+        if not isinstance(other, Path):
+            return NotImplemented
+        return self._path < other._path
+
+    def __le__(self, other: Any) -> bool:
+        if not isinstance(other, Path):
+            return NotImplemented
+        return self._path <= other._path
+
+    def __gt__(self, other: Any) -> bool:
+        if not isinstance(other, Path):
+            return NotImplemented
+        return self._path > other._path
+
+    def __ge__(self, other: Any) -> bool:
+        if not isinstance(other, Path):
+            return NotImplemented
+        return self._path >= other._path
+
+    @property
+    def name(self) -> str:
+        return self._path.name
+
+    @property
+    def suffix(self) -> str:
+        return self._path.suffix
+
+    @property
+    def suffixes(self) -> List[str]:
+        return self._path.suffixes
+
+    @property
+    def stem(self) -> str:
+        return self._path.stem
+
+    def with_name(self, name: str) -> "Path[H]":
+        return Path(self._host, self._path.with_name(name))
+
+    def with_stem(self, stem: str) -> "Path[H]":
+        return Path(self._host, self._path.with_name(stem + self._path.suffix))
```

Comment:
Why is this not `return Path(self._host, self._path.with_stem(stem))`?

Reply:
`pathlib.PurePath.with_stem()` was only introduced in Python 3.9.  As I try to keep tbot compatible back to 3.6 (for now), I chose to simply implement the method like it is implemented in `pathlib`, based on `with_name()` which already existed back then.

The same goes for `is_relative_to()`, btw.  The implementation here is copied from `pathlib` because it is a 3.9 feature, but trivial to implement ontop of existing methods.

Reply:
Oh, okay, totally understandable then! I would have added a comment explaining this rationale. Others will ask the same question.